go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -410,18 +422,86 @@ func (c *Controller) syncService(key string) error {
 			return fmt.Errorf("failed to ensure service %s load balancers: %w", key, err)
 		}
 
-		c.alreadyAppliedLock.Lock()
+		c.Lock()
 		c.alreadyApplied[key] = lbs
-		c.alreadyAppliedLock.Unlock()
+		c.Unlock()
 	}
 
 	c.repair.serviceSynced(key)
 	return nil
 }
 
+// c.nodeINfos, c.nodeIPv4Template and c.nodeIPv6Template are written by the node
+// tracker with the controller lock taken.  Make a copy before using them
+func (c *Controller) getNodeInfos() (nodeInfos []nodeInfo, nodeIPv4Template *Template, nodeIPv6Template *Template) {
+	c.Lock()
+	defer c.Unlock()
+
+	nodeInfos = make([]nodeInfo, 0, len(c.nodeInfos))
+	for _, nodeInfo := range c.nodeInfos {
+		nodeInfos = append(nodeInfos, nodeInfo)
+	}
+	nodeIPv4Template = c.nodeIPv4Template.deepCopy()
+	nodeIPv6Template = c.nodeIPv6Template.deepCopy()
```

Comment:
Yes, I considered locking wherever needed but we need the node info in a few places while processing a service change so I thought it might make more sense to have a consistent nodeInfo/nodeTemplate information for a given service change event.  We can indeed change it later if you think it's better.

Note: I still didn't squash the nodeIP -> svc controller change into its parent commit but I can do it once we agree on this approach.

Reply:
What I mean to say is just lock at the start and unlock at the end of processing a service instead of wherever needed. So at the end of the day just locking in two places:

- When processing a service
- When updating the nodeinfos

You could use a RWLock and read lock for service processing to have no contention between the workers, and write lock when updating the node info.

More contention but less deep copying and simpler code.

Reply:
Ok, I changed it to use locks.  I had to use two locks though.  We can't have a single RWMutex for both the alreadyApplied map and the node info.  That's because we need to write to the alreadyApplied map when processing a service.  I did change the alreadyAppliedLock to RWLock though.